<a href="https://colab.research.google.com/github/LuisHiram99/trabajos_PLN/blob/main/Tareas/Tarea%2002.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Tarea 2</h1>

<h3>Identificando la categoría de un texto usando unigramas y bigramas</h3>

# Dataset

In [1]:
!gdown 14cbj9FEc4YePGJg17LbjVSP42x6P960U

Downloading...
From: https://drive.google.com/uc?id=14cbj9FEc4YePGJg17LbjVSP42x6P960U
To: /content/two_categories_dataset.csv
100% 88.1M/88.1M [00:02<00:00, 38.1MB/s]


In [2]:
import pandas as pd

df = pd.read_csv('/content/two_categories_dataset.csv',index_col=0)
df.head()

text  tag
0  el ministerio español de ciencia e innovación ...    1
1  nos vamos profundamente emocionados de la hosp...    1
2  Lugar muy acogedor, tranquilo, menú a buen pre...    0
3  varios miembros del gobierno, entre ellos, el ...    1
4  Buena ubicación, a 15 minutos andando del cent...    0

In [4]:
df

text  tag
0      el ministerio español de ciencia e innovación ...    1
1      nos vamos profundamente emocionados de la hosp...    1
2      Lugar muy acogedor, tranquilo, menú a buen pre...    0
3      varios miembros del gobierno, entre ellos, el ...    1
4      Buena ubicación, a 15 minutos andando del cent...    0
...                                                  ...  ...
36339  la constructora ohl habría gastado hasta más d...    1
36340  El hotel tiene una situación inmejorable para ...    0
36341  Estuvimos pasando una noche en el Palacete por...    0
36342  Situado en zona peatonal en las calles estrech...    0
36343  Fui por primera vez a Jerez de la Frontera en ...    0

[36344 rows x 2 columns]

# Tareas



1.   Separar los textos por categoria
2.   Analisis exploratio / preprocesamiento
3.   K palabras mas frecuentes en cada etiqueta
4.   k biagramas mas frec en cada etiqueta
5.  Estrategia para asignar etiqueta



In [5]:
#1:
df_tag_0 = df[df['tag']==0]
df_tag_1 = df[df['tag']==1]

In [8]:
df_tag_0.columns

Index(['text', 'tag'], dtype='object')

In [17]:
texto = df_tag_0["text"][2]

In [19]:
import nltk
from nltk import bigrams, trigrams
from nltk.corpus import reuters
from collections import defaultdict
from string import punctuation

nltk.download('punkt_tab')
nltk.download('stopwords')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [36]:
stopword_es = nltk.corpus.stopwords.words('spanish')
stopword_es

['de',
 'la',
 'que',
 'el',
 'en',
 'y',
 'a',
 'los',
 'del',
 'se',
 'las',
 'por',
 'un',
 'para',
 'con',
 'no',
 'una',
 'su',
 'al',
 'lo',
 'como',
 'más',
 'pero',
 'sus',
 'le',
 'ya',
 'o',
 'este',
 'sí',
 'porque',
 'esta',
 'entre',
 'cuando',
 'muy',
 'sin',
 'sobre',
 'también',
 'me',
 'hasta',
 'hay',
 'donde',
 'quien',
 'desde',
 'todo',
 'nos',
 'durante',
 'todos',
 'uno',
 'les',
 'ni',
 'contra',
 'otros',
 'ese',
 'eso',
 'ante',
 'ellos',
 'e',
 'esto',
 'mí',
 'antes',
 'algunos',
 'qué',
 'unos',
 'yo',
 'otro',
 'otras',
 'otra',
 'él',
 'tanto',
 'esa',
 'estos',
 'mucho',
 'quienes',
 'nada',
 'muchos',
 'cual',
 'poco',
 'ella',
 'estar',
 'estas',
 'algunas',
 'algo',
 'nosotros',
 'mi',
 'mis',
 'tú',
 'te',
 'ti',
 'tu',
 'tus',
 'ellas',
 'nosotras',
 'vosotros',
 'vosotras',
 'os',
 'mío',
 'mía',
 'míos',
 'mías',
 'tuyo',
 'tuya',
 'tuyos',
 'tuyas',
 'suyo',
 'suya',
 'suyos',
 'suyas',
 'nuestro',
 'nuestra',
 'nuestros',
 'nuestras',
 'vuestro'

In [37]:
punctuations = list(punctuation)

words = [t for t in nltk.word_tokenize(texto) if t not in punctuations and t not in stopword_es]

words

['Lugar',
 'acogedor',
 'tranquilo',
 'menú',
 'buen',
 'precio',
 'comida',
 'típica',
 'Sierra',
 'Buena',
 'atención',
 'cliente',
 'zona',
 'parquing',
 'amplia',
 'buenas',
 'vistas',
 'Pontones',
 'Ideal',
 'hacer',
 'visita',
 'prisas']

In [39]:
from collections import Counter

frequent_words = Counter(words)
frequent_words.most_common(10)


[('Lugar', 1),
 ('acogedor', 1),
 ('tranquilo', 1),
 ('menú', 1),
 ('buen', 1),
 ('precio', 1),
 ('comida', 1),
 ('típica', 1),
 ('Sierra', 1),
 ('Buena', 1)]

In [43]:
df_tag_0.reset_index()
df_tag_0["text"]

0        Lugar muy acogedor, tranquilo, menú a buen pre...
1        Buena ubicación, a 15 minutos andando del cent...
2        Lo pasamos genial,  todo maravilloso,  la limp...
3        Servicio inmejorable! Excelente desayuno! Ropa...
4        Hotel sencillo, limpio y tranquilo cuyo mayor ...
                               ...                        
18167    El Hostal Delfin Verde esta en un punto incomp...
18168    El hotel tiene una situación inmejorable para ...
18169    Estuvimos pasando una noche en el Palacete por...
18170    Situado en zona peatonal en las calles estrech...
18171    Fui por primera vez a Jerez de la Frontera en ...
Name: text, Length: 18172, dtype: object

In [60]:
df_tag_1 = df_tag_1["text"].reset_index()

In [51]:
def get_all_text(df):
  list_of_tokens = []
  for i in range(0,df["text"].size):
    texto = df["text"][i]
    words = [t.lower() for t in nltk.word_tokenize(texto) if t not in punctuations and t not in stopword_es]
    list_of_tokens.extend(words)
  return list_of_tokens

In [53]:
all_text_tokens = get_all_text(df_tag_0)

In [54]:
all_text_freq = Counter(all_text_tokens)
all_text_freq.most_common(10)

[('hotel', 12049),
 ('El', 10118),
 ('La', 8417),
 ('habitación', 7659),
 ('bien', 7168),
 ('personal', 5390),
 ('habitaciones', 4843),
 ('desayuno', 4061),
 ('si', 4013),
 ('centro', 3682)]

In [61]:
all_text_tokens = get_all_text(df_tag_1)
all_text_freq = Counter(all_text_tokens)
all_text_freq.most_common(10)

[('``', 115217),
 ("''", 107887),
 ('»', 57028),
 ('gobierno', 35467),
 ('personas', 24455),
 ('si', 23960),
 ('horas', 17934),
 ('podemos', 17899),
 ('madrid', 17445),
 ('años', 17393)]

In [63]:
df_tag_1["text"][0]

'el ministerio español de ciencia e innovación tiene registrados diez proyectos de investigación de vacunas frente al coronavirus en todo el territorio nacional. si bien existen propuestas en otros países del mundo en esta línea, según la asociación española de vacunología (aev), la mayoría se encuentra en estadios muy incipientes. de las diez que se están investigando en españa encabezan la lista las tres del consejo superior de investigaciones científicas (csic), en concreto dos que se desarrollan en el centro nacional de biotecnología (cnb) el proyecto liderado por los científicos luis enjuanes e isabel sola, y el de juan garcía arriaza y el tercero, en el centro de investigaciones biológicas (cib), apellidado recientemente margarita salas en honor a esta insigne investigadora de proyección internacional, y capitaneado por los profesores vicente larraga y mariano esteban. el virólogo enjuanes prevé que su vacuna estará lista a principios del año que viene empleamos dos métodos de ad

[('hotel', 12049),
 ('El', 10118),
 ('La', 8417),
 ('habitación', 7659),
 ('bien', 7168),
 ('personal', 5390),
 ('habitaciones', 4843),
 ('desayuno', 4061),
 ('si', 4013),
 ('centro', 3682)]


[('``', 115217),
 ("''", 107887),
 ('»', 57028),
 ('gobierno', 35467),
 ('personas', 24455),
 ('si', 23960),
 ('horas', 17934),
 ('podemos', 17899),
 ('madrid', 17445),
 ('años', 17393)]